### PR TITLE
Add tests to kill escaped mutants (#529)

### PR DESCRIPTION
**Pull request by dbuhonov**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | part of #529

Covered Code MSI: 98%

### CODE REVIEW

File: tests/ActiveRecordTest.php [hunk 1]
```diff
@@ -436,6 +438,18 @@ public function testResetNotSavedRelation(): void
...
+    public function testResetRelation(): void
+    {
+        $customer = Customer::query()->findByPk(2);
+        $customer->getOrders();
+
+        $this->assertStringContainsString('orders', serialize($customer));
+
+        $customer->resetRelation('orders');
+
+        $this->assertStringNotContainsString('orders', serialize($customer));
```
[the remark this answers is not in the document]
Comment (dbuhonov) [reply tests/ActiveRecordTest.php]: Add this

File: infection.json.dist [hunk 2]
```diff
@@ -13,6 +13,45 @@
...
+            ],
+            "ignoreSourceCodeByRegex": [
+                "protected function upsertInternal\\(\\?array \\$insertProperties = null, array\\|bool \\$updateProperties = true\\): void"
+            ]
+        },
+        "UnwrapArrayValues": {
+            "ignore": [
+                "Yiisoft\\ActiveRecord\\Internal\\JoinsWithBuilder::build"
+            ]
+        }
```
[the remark this answers is not in the document]
Comment (dbuhonov) [reply infection.json.dist]: These exclusions are for survivors that are not obvious from the test diff itself, so I moved them into infection.json.dist intentionally to make them explicit and visible rather than leaving them hidden in the report only.

File: tests/ActiveRecordTest.php [hunk 3]
```diff
@@ -436,6 +449,41 @@ public function testResetNotSavedRelation(): void
...
+        $this->assertArrayNotHasKey('orders', $customer->relatedRecords());
+    }
+
+    public function testResetRelationRemovesRelationFromDependencies(): void
+    {
+        $order = OrderWithCustomerProfileViaCustomerRelation::query()->findByPk(1);
+        $order->getCustomerProfileViaCustomer();
+
+        $reflection = new ReflectionProperty(AbstractActiveRecord::class, 'relationsDependencies');
+        $dependencies = $reflection->getValue($order);
```
[the remark this answers is not in the document]
Comment (dbuhonov) [reply tests/ActiveRecordTest.php]: Not applicable for this project. The package supports PHP 8.1+ only, and since PHP 8.1 ReflectionProperty::getValue() can read private properties without calling setAccessible(true). This would be a valid concern for PHP 7.4, but that version is outside our supported range.

File: tests/Driver/Pgsql/ActiveQueryTest.php [hunk 4]
```diff
@@ -21,6 +28,70 @@ public function testBit(): void
...
+            ['id' => 1],
+        ]);
+
+        $where = $query->getWhere();
+
+        $this->assertInstanceOf(ArrayOverlaps::class, $where);
+        $this->assertSame('array_item_ids', $where->column);
+        $this->assertInstanceOf(ArrayValue::class, $where->values);
+        $this->assertSame([1], $where->values->value);
+    }
```
Comment (Tigrov) [inline tests/Driver/Pgsql/ActiveQueryTest.php]: Doesn't it duplicate the next test?

File: tests/Stubs/ActiveRecord/CustomerWithDeleteInternalOverride.php [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\ActiveRecord\Tests\Stubs\ActiveRecord;
+
+final class CustomerWithDeleteInternalOverride extends Customer
```
Comment (Tigrov) [inline tests/Stubs/ActiveRecord/CustomerWithDeleteInternalOverride.php]: Where is this class used?

File: tests/Stubs/ActiveRecord/CustomerWithOverriddenRelationQuery.php [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+declare(strict_types=1);
+
+namespace Yiisoft\ActiveRecord\Tests\Stubs\ActiveRecord;
+
+use Yiisoft\ActiveRecord\ActiveQueryInterface;
+use Yiisoft\ActiveRecord\ActiveRecordInterface;
+use Yiisoft\ActiveRecord\Tests\Stubs\ActiveQuery\AlternativeActiveQuery;
+
+final class CustomerWithOverriddenRelationQuery extends Customer
```
Comment (Tigrov) [inline tests/Stubs/ActiveRecord/CustomerWithOverriddenRelationQuery.php]: Where is this class used?

File: tests/Stubs/ActiveRecord/CustomerWithRefreshInternalOverride.php [hunk 7]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\ActiveRecord\Tests\Stubs\ActiveRecord;
+
+use Yiisoft\ActiveRecord\ActiveRecordInterface;
+
+final class CustomerWithRefreshInternalOverride extends Customer
```
Comment (Tigrov) [inline tests/Stubs/ActiveRecord/CustomerWithRefreshInternalOverride.php]: Where is this class used?

File: tests/Stubs/ActiveRecord/CustomerWithUpdateInternalOverride.php [hunk 8]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\ActiveRecord\Tests\Stubs\ActiveRecord;
+
+final class CustomerWithUpdateInternalOverride extends Customer
```
Comment (Tigrov) [inline tests/Stubs/ActiveRecord/CustomerWithUpdateInternalOverride.php]: Where is this class used?

File: tests/Stubs/ActiveRecord/EmployeeWithPrototypeDossierRelation.php [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+declare(strict_types=1);
+
+namespace Yiisoft\ActiveRecord\Tests\Stubs\ActiveRecord;
+
+use Yiisoft\ActiveRecord\ActiveQueryInterface;
+use Yiisoft\ActiveRecord\ActiveRecord;
+use Yiisoft\ActiveRecord\ActiveRecordInterface;
+
+final class EmployeeWithPrototypeDossierRelation extends ActiveRecord
```
Comment (Tigrov) [inline tests/Stubs/ActiveRecord/EmployeeWithPrototypeDossierRelation.php]: Where is this class used?

File: tests/Stubs/ActiveRecord/CustomerWithOverriddenRelationQuery.php (hunk 6 above)
Comment (dbuhonov) [reply tests/Stubs/ActiveRecord/CustomerWithOverriddenRelationQuery.php]: It is used testCreateRelationQueryCanBeOverridden

File: tests/Stubs/ActiveRecord/CustomerWithRefreshInternalOverride.php (hunk 7 above)
Comment (dbuhonov) [reply tests/Stubs/ActiveRecord/CustomerWithRefreshInternalOverride.php]: it is used testRefreshInternalCanBeOverridden

File: tests/Stubs/ActiveRecord/CustomerWithDeleteInternalOverride.php (hunk 5 above)
Comment (dbuhonov) [reply tests/Stubs/ActiveRecord/CustomerWithDeleteInternalOverride.php]: It is used testDeleteInternalCanBeOverridden

File: tests/Stubs/ActiveRecord/CustomerWithUpdateInternalOverride.php (hunk 8 above)
Comment (dbuhonov) [reply tests/Stubs/ActiveRecord/CustomerWithUpdateInternalOverride.php]: It is used testUpdateInternalCanBeOverridden

File: tests/Stubs/ActiveRecord/EmployeeWithPrototypeDossierRelation.php (hunk 9 above)
Comment (dbuhonov) [reply tests/Stubs/ActiveRecord/EmployeeWithPrototypeDossierRelation.php]: It is used testLinkExistingRecordToNewWithStrictPrimaryKeyValidationUsesLinkValues

File: tests/Driver/Pgsql/ActiveQueryTest.php (hunk 4 above)
Comment (dbuhonov) [reply tests/Driver/Pgsql/ActiveQueryTest.php]: Merged these checks into a single test: testModelRelationFilterUsesArrayOverlapsWithArrayValueAndColumnSchemaForArrayColumns, since both were covering the same ColumnType::ARRAY

File: CHANGELOG.md [hunk 10]
```diff
@@ -2,7 +2,7 @@
 
 ## 1.0.3 under development
 
-- no changes in this release.
+- Test #529: Increase MSI to 100% (@dbuhonov)
```
Comment (vjik) [inline CHANGELOG.md]: These changes don't touch package code. Line in changelog is not needed.

Comment (dbuhonov) [reply CHANGELOG.md]: Revert

**Comment by dbuhonov** [comment]
Up review pls

File: tests/ActiveRecordTest.php [hunk 11]
```diff
@@ -1664,6 +1914,64 @@ public function testLinkExistingRecordToNewWithSharedPrimaryKey(): void
...
+        $dossierPrototype = new CompositePrimaryKeyDossier();
+
+        $employeePrototype = new EmployeeWithPrototypeDossierRelation($dossierPrototype);
+
+        $employee = (new ActiveQuery(clone $employeePrototype))->findByPk([2, 2]);
+        $this->assertNotNull($employee);
+
+        $dossier = clone $dossierPrototype;
+        $dossier->set('summary', 'Strict primary key validation');
+        if ($this->db()->getDriverName() !== 'sqlsrv') {
```
Comment (Tigrov) [inline tests/ActiveRecordTest.php]: Why skip `sqlsrv` here?

File: tests/ActiveRecordTest.php [hunk 12]
```diff
@@ -1664,6 +1914,64 @@ public function testLinkExistingRecordToNewWithSharedPrimaryKey(): void
...
+            $dossier->set('id', 100);
+        }
+
+        $employee->link('dossier', $dossier);
+
+        $this->assertSame(2, $dossier->get('department_id'));
+        $this->assertSame(2, $dossier->get('employee_id'));
+        $this->assertNotNull($dossier->get('id'));
```
Comment (Tigrov) [inline tests/ActiveRecordTest.php]: Should it be equal to `100`?

File: tests/ActiveRecordTest.php [hunk 13]
```diff
@@ -1730,6 +2038,32 @@ public function testMarkPropertyChanged(): void
...
 
+    public function testMarkPropertyChangedWithEmptyNameDoesNotModifyOldValues(): void
+    {
+        $customer = new Customer();
+        $customer->assignOldValues(['' => 'sentinel', 'name' => 'user1']);
+
+        $customer->markPropertyChanged('');
+
+        $this->assertSame(['' => 'sentinel', 'name' => 'user1'], $customer->oldValues());
+    }
```
Comment (Tigrov) [inline tests/ActiveRecordTest.php]: Perhaps this is a bug and `$customer->markPropertyChanged('');` should change it.

File: tests/ActiveRecordTest.php [hunk 14]
```diff
@@ -1931,6 +2305,82 @@ public function testUpdateCountersThrowsExceptionForNewRecord(): void
...
+    {
+        $this->reloadFixtureAfterTest();
+
+        $customer = CustomerWithUpdateInternalOverride::query()->findByPk(1);
+
+        $customer->setName('should-not-hit-db');
+
+        $this->assertSame(91, $customer->update());
+        $this->assertSame('user1', Customer::query()->findByPk(1)->getName());
+    }
```
Comment (Tigrov) [inline tests/ActiveRecordTest.php]: This is where the fundamentals of OOP (inheritance) are tested. This looks redundant, but if it is necessary, perhaps it would be better to create a separate class in which to define all methods which can be overridden.

File: tests/ArrayAccessTraitTest.php [hunk 15]
```diff
@@ -23,13 +25,30 @@ public function testOffsetExists(): void
...
     }
 
+    public function testOffsetExistsDoesNotTreatPropertyAsRelation(): void
+    {
+        $model = new CustomerArrayAccessModel();
+        $model->name = 'test';
+
+        $this->assertTrue(isset($model['name']));
+        $this->assertFalse($model->isRelationPopulated('name'));
+    }
```
Comment (Tigrov) [inline tests/ArrayAccessTraitTest.php:34]: The test looks like `testOffsetGetReturnsPropertyValue()`
```suggestion
```

File: tests/ArrayAccessTraitTest.php [hunk 16]
```diff
@@ -23,13 +25,30 @@ public function testOffsetExists(): void
...
         $this->assertTrue(isset($model['profile']));
     }
 
+    public function testOffsetExistsWithMagicProperty(): void
+    {
+        $model = new CategoryWithArrayAccess();
+        $model['name'] = 'magic';
+
+        $this->assertTrue(isset($model['name']));
+    }
```
Comment (Tigrov) [inline tests/ArrayAccessTraitTest.php]: The test looks like `testOffsetGetWithMagicProperty()`
```suggestion
```

File: tests/ArrayAccessTraitTest.php [hunk 17]
```diff
@@ -41,13 +60,30 @@ public function testOffsetGet(): void
...
 
         $this->assertInstanceOf(Profile::class, $model['profile']);
     }
 
+    public function testOffsetGetWithMagicProperty(): void
+    {
+        $model = new CategoryWithArrayAccess();
+        $model['name'] = 'magic';
+
+        $this->assertSame('magic', $model['name']);
```
Comment (Tigrov) [inline tests/ArrayAccessTraitTest.php]: Move this to `testOffsetSetWithProperty()`

File: tests/ArrayAccessTraitTest.php [hunk 18]
```diff
@@ -41,13 +60,30 @@ public function testOffsetGet(): void
         $this->assertSame('custom value', $model['customProperty']);
     }
 
+    public function testOffsetGetReturnsPropertyValue(): void
+    {
+        $model = new CustomerArrayAccessModel();
+        $model->name = 'property value';
+
+        $this->assertSame('property value', $model['name']);
+        $this->assertFalse($model->isRelationPopulated('name'));
```
Comment (Tigrov) [inline tests/ArrayAccessTraitTest.php:50]: Move this to `testOffsetGet()`

File: tests/ArrayAccessTraitTest.php [hunk 19]
```diff
@@ -124,6 +161,21 @@ public function testOffsetUnsetWithObjectProperty(): void
         $this->assertTrue(!isset($model->customProperty));
     }
 
+    public function testOffsetUnsetWithPropertyResetsDependentRelation(): void
+    {
+        $this->reloadFixtureAfterTest();
```
Comment (Tigrov) [inline tests/ArrayAccessTraitTest.php]: No need reload fixtures, if no changes in db
```suggestion
```

File: tests/ArrayAccessTraitTest.php [hunk 20]
```diff
@@ -135,4 +187,28 @@ public function testOffsetUnsetWithRelation(): void
...
+
+        $model['name'] = 'magic';
+        $model->populateRelation('name', new Profile());
+
+        unset($model['name']);
+
+        $this->assertNull($model->get('name'));
+        $this->assertTrue($model->isRelationPopulated('name'));
+        $this->assertInstanceOf(Profile::class, $model->relation('name'));
+    }
```
Comment (Tigrov) [inline tests/ArrayAccessTraitTest.php]: Doesn't it look like a bug?
Either relationships cannot have the same name as properties, or `unset()` must remove both values.

File: tests/EventsTraitTest.php [hunk 21]
```diff
@@ -154,6 +170,25 @@ static function (object $event): void {
...
+                    }
+                },
+            ),
+        );
+
+        $model = CategoryEventsModel::query()->findByPk(1);
+
+        $this->assertSame(0, $model->delete());
+        $this->assertNotNull(CategoryEventsModel::query()->findByPk(1));
+    }
```
Comment (Tigrov) [inline tests/EventsTraitTest.php]: Use `SoftDelete` event handler to test `preventDefault()` on delete.
Or `Order` stub class which uses `SoftDelete`

Perhaps double of `testSoftDeleteBeforeDeleteUsesCustomValueAndPreventsDefault()`

File: tests/EventsTraitTest.php [hunk 22]
```diff
@@ -251,4 +286,240 @@ static function (object $event): void {
...
+            ),
+        );
+        EventDispatcherProvider::set(
+            CustomerEventsModel::class,
+            new SimpleEventDispatcher(
+                static function (object $event) use (&$triggeredEvents): void {
+                    $triggeredEvents[] = $event::class;
+                },
+            ),
+        );
```
Comment (Tigrov) [inline tests/EventsTraitTest.php:321]: Why need to set it twice?

File: tests/EventsTraitTest.php [hunk 23]
```diff
@@ -251,4 +286,240 @@ static function (object $event): void {
...
+
+        $handler = new DefaultValueOnInsert('Vasya', 'name');
+        $eventHandlers = $handler->getEventHandlers();
+        $beforeUpsert = $eventHandlers[BeforeUpsert::class];
+        $beforeUpsert($event);
+
+        $this->assertSame(
+            [0 => 'id', 1 => 'name', 'name' => 'Vasya'],
+            $insertProperties,
+        );
```
Comment (Tigrov) [inline tests/EventsTraitTest.php]: Doesn't it looks like a bug? Perhaps the correct result should be `[0 => 'id', 'name' => 'Vasya'],`

File: tests/EventsTraitTest.php [hunk 24]
```diff
@@ -251,4 +286,240 @@ static function (object $event): void {
...
+        $insertProperties = ['id' => 1];
+        $updateProperties = false;
+        $event = new BeforeUpsert($model, $insertProperties, $updateProperties);
+
+        $handler = new SetValueOnUpdate('Updated', 'name');
+        $eventHandlers = $handler->getEventHandlers();
+        $beforeUpsert = $eventHandlers[BeforeUpsert::class];
+        $beforeUpsert($event);
+
+        $this->assertSame(['name' => 'Updated'], $updateProperties);
```
Comment (Tigrov) [inline tests/EventsTraitTest.php]: Doesn't it look like a bug?
If upsert called with `$updateProperties = false`, should it update values specified by `SetValueOnUpdate`?

File: tests/EventsTraitTest.php [hunk 25]
```diff
@@ -251,4 +286,240 @@ static function (object $event): void {
...
+        $event = new BeforeDelete($order);
+        $handler = new SoftDelete(new DateTimeImmutable('2022-03-04 05:06:07'), 'deleted_at');
+        $eventHandlers = $handler->getEventHandlers();
+        $beforeDelete = $eventHandlers[BeforeDelete::class];
+        $beforeDelete($event);
+
+        $this->assertTrue($event->isDefaultPrevented());
+        $this->assertNull($event->getReturnValue());
+        $this->assertNull(self::db()->select('deleted_at')->from('{{order}}')->where(['id' => 1])->scalar());
+    }
```
Comment (Tigrov) [inline tests/EventsTraitTest.php]: This is a bug.

File: tests/MagicActiveRecordTest.php [hunk 26]
```diff
@@ -235,6 +236,19 @@ public function testPopulateRecordCallWhenQueryingOnParentClass(): void
...
+        $this->reloadFixtureAfterTest();
+
+        $record = new SetValueOnUpdateAr();
+        $record->id = 99;
+        $record->name = 'Test';
+
+        $record->save();
+
+        $this->assertSame('Test', $record->name);
+    }
```
Comment (Tigrov) [inline tests/MagicActiveRecordTest.php]: `SetValueOnUpdateAr` class is not magic. The test should be moved to `ActiveRecordTest.php`

File: tests/ActiveQueryTest.php [hunk 27]
```diff
@@ -73,7 +107,315 @@ public function testPrepare(): void
...
+
+    public function testArArrayHelperIndexCastsFloatKeyToString(): void
+    {
+        $indexed = ArArrayHelper::index([
+            ['id' => 1.5, 'name' => 'float-key'],
+        ], 'id');
+
+        $this->assertArrayHasKey('1.5', $indexed);
+        $this->assertSame('float-key', $indexed['1.5']['name']);
+    }
```
Comment (Tigrov) [inline tests/ActiveQueryTest.php:145]: This tests the basics of PHP, if you pass `float` as a key to an array, the key will become `string`.

File: tests/ActiveQueryTest.php [hunk 28]
```diff
@@ -73,7 +107,315 @@ public function testPrepare(): void
...
+        $this->assertCount(2, $models);
+        $this->assertSame([1, 2], array_column($models, 'item_id'));
+    }
+
+    public function testGetTableNameAndAliasDoesNotTrimTrailingGarbage(): void
+    {
+        [$tableName, $alias] = TableNameAndAliasResolver::resolve(Customer::query()->from(['customer c!']));
+
+        $this->assertSame('customer c!', $tableName);
+        $this->assertSame('customer c!', $alias);
```
Comment (Tigrov) [inline tests/ActiveQueryTest.php]: Doesn't it looks like a bug?

File: tests/ActiveQueryTest.php [hunk 29]
```diff
@@ -73,7 +107,315 @@ public function testPrepare(): void
...
+    }
+
+    public function testGetTableNameAndAliasDoesNotExtractAliasFromSuffixAfterNewline(): void
+    {
+        $from = "ignored\ncustomer c";
+
+        [$tableName, $alias] = TableNameAndAliasResolver::resolve(Customer::query()->from([$from]));
+
+        $this->assertSame($from, $tableName);
+        $this->assertSame($from, $alias);
```
Comment (Tigrov) [inline tests/ActiveQueryTest.php]: Doesn't it look like a bug?

File: tests/ActiveQueryTest.php [hunk 30]
```diff
@@ -73,7 +107,315 @@ public function testPrepare(): void
...
+        $query = new CreateModelsExceptionOnEmptyRowsActiveQuery(Customer::class);
+
+        $this->assertSame([], $query->populate([]));
+    }
+
+    public function testRemoveDuplicatedRowsChecksPrimaryKeyPresenceInFirstRow(): void
+    {
+        $query = Customer::query();
+        $method = new ReflectionMethod(ActiveQuery::class, 'removeDuplicatedRows');
+        $method->setAccessible(true);
```
Comment (Tigrov) [inline tests/ActiveQueryTest.php]: This is not required
https://www.php.net/manual/en/reflectionmethod.setaccessible.php

> Note: As of PHP 8.1.0, calling this method has no effect; all methods are invokable by default.
```suggestion
```

File: tests/ActiveQueryTest.php [hunk 31]
```diff
@@ -73,7 +107,315 @@ public function testPrepare(): void
...
+            $query->populate([['id' => 1]]),
+        );
+    }
+
+    public function testGetPrimaryTableNameCanBeOverridden(): void
+    {
+        $query = new OverriddenPrimaryTableNameActiveQuery(Customer::class);
+
+        $this->assertSame(['{{profile}}' => '{{profile}}'], $query->getTablesUsedInFrom());
+    }
```
Comment (Tigrov) [inline tests/ActiveQueryTest.php]: This is where the fundamentals of OOP (inheritance) are tested. This looks redundant, but if it is necessary, perhaps it would be better to create a separate class in which to define all methods which can be overridden.

File: tests/ActiveQueryTest.php [hunk 32]
```diff
@@ -73,7 +107,315 @@ public function testPrepare(): void
...
+        ]);
+
+        $where = $query->getWhere();
+
+        $this->assertInstanceOf(In::class, $where);
+        $this->assertSame(
+            [['d.department_id' => 2, 'd.employee_id' => null]],
+            array_values($where->values),
+        );
+    }
```
Comment (Tigrov) [inline tests/ActiveQueryTest.php]: It looks like `testModelRelationFilterCompositeKeysFillMissingValuesWithNull()`

File: tests/EventsTraitTest.php (hunk 25 above)
Comment (Tigrov) [reply tests/EventsTraitTest.php]: Fixed by https://github.com/yiisoft/active-record/pull/558

File: tests/ArrayAccessTraitTest.php (hunk 15 above)
Comment (dbuhonov) [reply tests/ArrayAccessTraitTest.php:34]: Removed

File: tests/ActiveQueryTest.php (hunk 32 above)
Comment (dbuhonov) [reply tests/ActiveQueryTest.php]: Renamed it to make the distinction clearer.This one covers array models with an aliased primary table

File: tests/ActiveQueryTest.php (hunk 31 above)
Comment (dbuhonov) [reply tests/ActiveQueryTest.php]: Agreed, these tests dont add much beyond checking overridable methods. I removed them

File: tests/ActiveQueryTest.php (hunk 30 above)
Comment (dbuhonov) [reply tests/ActiveQueryTest.php]: Thx, removed

File: tests/ActiveQueryTest.php (hunk 29 above)
Comment (dbuhonov) [reply tests/ActiveQueryTest.php]: The initial idea was to ensure that the resolver doesn’t try to infer an alias from complex/multiline FROM strings and only extracts aliases from simple table expressions.
But I agree this case is artificial and the fallback result looks confusing, so I removed the test

File: tests/ActiveQueryTest.php (hunk 28 above)
Comment (dbuhonov) [reply tests/ActiveQueryTest.php]: I agree that returning the whole string as both table name and alias looks confusing and doesnt document useful behavior, so I removed the test

File: tests/ActiveQueryTest.php (hunk 27 above)
Comment (dbuhonov) [reply tests/ActiveQueryTest.php:145]: PHP actually casts float array keys to int, not string, so `1.5` would become `1`
This test covers the explicit string cast in ArArrayHelper::index(), which preserves `1.5` as `'1.5'`

File: tests/MagicActiveRecordTest.php (hunk 26 above)
Comment (dbuhonov) [reply tests/MagicActiveRecordTest.php]: Agreed. Moved it to ActiveRecordTest and renamed the test to testSetValueOnUpdateSaveNewRecordDoesNotExecuteUpdate.

File: tests/EventsTraitTest.php (hunk 25 above)
Comment (dbuhonov) [reply tests/EventsTraitTest.php]: Agreed, bug was fixed by #558, this test is no needed. Removed

File: tests/EventsTraitTest.php (hunk 24 above)
Comment (dbuhonov) [reply tests/EventsTraitTest.php]: Agreed, this looks like a bug. `updateProperties: false` should keep updates disabled. Test removed to avoid recording this behavior

File: tests/EventsTraitTest.php (hunk 23 above)
Comment (dbuhonov) [reply tests/EventsTraitTest.php]: Agreed, this bug. The same property should not be present both as a numeric property name and as a keyed value. Removed test

File: tests/EventsTraitTest.php (hunk 22 above)
Comment (dbuhonov) [reply tests/EventsTraitTest.php:321]: Because EventDispatcherProvider stores dispatchers per target class, and this test dispatches events from both CategoryEventsModel and CustomerEventsModel

File: tests/EventsTraitTest.php (hunk 21 above)
Comment (dbuhonov) [reply tests/EventsTraitTest.php]: And testDeleteWithEventPrevention() checks the flow of the delete() method. Renamed for transparencytestDeleteReturnsZeroAndSkipsDeletionWhenBeforeDeletePreventsDefault

File: tests/ArrayAccessTraitTest.php (hunk 20 above)
Comment (dbuhonov) [reply tests/ArrayAccessTraitTest.php]: I dont know, but it looks doubtful. I removed the test so as not to reinforce this behavior. But it seems to me that both should be reset

File: tests/ArrayAccessTraitTest.php (hunk 19 above)
Comment (dbuhonov) [reply tests/ArrayAccessTraitTest.php]: Removed

File: tests/ArrayAccessTraitTest.php (hunk 18 above)
Comment (dbuhonov) [reply tests/ArrayAccessTraitTest.php:50]: Done, moved this assertion into testOffsetGet()

File: tests/ArrayAccessTraitTest.php (hunk 16 above)
Comment (dbuhonov) [reply tests/ArrayAccessTraitTest.php]: Removed testOffsetExistsWithMagicProperty

File: tests/ActiveRecordTest.php (hunk 14 above)
Comment (dbuhonov) [reply tests/ActiveRecordTest.php]: Agreed, these tests are not necessary for this PR. Removed them

File: tests/ActiveRecordTest.php (hunk 13 above)
Comment (dbuhonov) [reply tests/ActiveRecordTest.php]: Agreed, this edge case is questionable. An empty property name should either be handled consistently or rejected
Since this PR is test-only, I removed test

File: tests/ActiveRecordTest.php (hunk 11 above)
Comment (dbuhonov) [reply tests/ActiveRecordTest.php]: SQL Server uses IDENTITY for `dossier.id`, so inserting an explicit `id` fails unless IDENTITY_INSERT is enabled otherwise the test fails

File: tests/ActiveRecordTest.php (hunk 12 above)
Comment (dbuhonov) [reply tests/ActiveRecordTest.php]: Only for drivers where we set `id` explicitly. For sqlsrv, `dossier.id` is an IDENTITY column, so we don’t set it manually and the database generates the value.
Added  conditional assertion for this.

File: tests/EventsTraitTest.php (hunk 22 above)
Comment (Tigrov) [reply tests/EventsTraitTest.php:321]: Oh, sorry, I didn't notice that the class names were different.

File: tests/ActiveRecordTest.php (hunk 11 above)
Comment (Tigrov) [reply tests/ActiveRecordTest.php]: Looks like test fixtures are not consistent. In each DBMS `dossier.id` should be marked as `AUTO_INCREMENT` / `IDENTITY` or as not.

File: tests/ActiveRecordTest.php (hunk 14 above)
Comment (Tigrov) [reply tests/ActiveRecordTest.php]: Also needs to remove classes created for the tests

File: tests/ArrayAccessTraitTest.php (hunk 20 above)
Comment (Tigrov) [reply tests/ArrayAccessTraitTest.php]: Moved to issue #552

File: tests/EventsTraitTest.php (hunk 23 above)
Comment (Tigrov) [reply tests/EventsTraitTest.php]: Moved to issue #559

File: tests/EventsTraitTest.php (hunk 24 above)
Comment (Tigrov) [reply tests/EventsTraitTest.php]: Moved to issue #553

File: tests/ActiveQueryTest.php (hunk 31 above)
Comment (Tigrov) [reply tests/ActiveQueryTest.php]: Also needs to remove classes created for the tests

File: tests/ActiveQueryTest.php (hunk 28 above)
Comment (Tigrov) [reply tests/ActiveQueryTest.php]: Moved this to issue #560

File: tests/ActiveQueryTest.php (hunk 29 above)
Comment (Tigrov) [reply tests/ActiveQueryTest.php]: Moved this to issue #560

File: tests/ActiveRecordTest.php (hunk 14 above)
Comment (dbuhonov) [reply tests/ActiveRecordTest.php]: Done

File: tests/ActiveQueryTest.php (hunk 31 above)
Comment (dbuhonov) [reply tests/ActiveQueryTest.php]: Done, removed all unused files

File: tests/ActiveRecordTest.php (hunk 11 above)
Comment (dbuhonov) [reply tests/ActiveRecordTest.php]: For now I removed testLinkExistingRecordToNewWithStrictPrimaryKeyValidationUsesLinkValues and testLinkNewRecordToExistingWithCustomCompositeSharedPrimaryKey. Id rather handle that in a separate

File: src/Event/Handler/SoftDelete.php
Comment (Tigrov) [inline src/Event/Handler/SoftDelete.php:1]: https://github.com/yiisoft/active-record/pull/545/changes#diff-ffae17165f1ebb2db4b7d7d3064d7e9b101ca81041c86299767c6d87d491b43bR22-R87

These changes should be moved in separate PR if they still actual.

**Review by Tigrov — CHANGES_REQUESTED** [review]

File: src/Event/Handler/SoftDelete.php
Comment (dbuhonov) [reply src/Event/Handler/SoftDelete.php:1]: I rolled back the changes to close them after

**Review by Tigrov — APPROVED** [review]
Could you merge changes from master?

**Comment by dbuhonov** [comment]
> Could you merge changes from master?

Yes, I just resolve conflicts)

**Comment by dbuhonov** [comment]
@Tigrov Fixed the conflict. The checks that failed are not my fault.

File: tests/ActiveQueryTest.php [hunk 33]
```diff
@@ -71,7 +102,258 @@ public function testPrepare(): void
...
+    {
+        $query = new CreateModelsExceptionOnEmptyRowsActiveQuery(Customer::class);
+
+        $this->assertSame([], $query->populate([]));
+    }
+
+    public function testRemoveDuplicatedRowsChecksPrimaryKeyPresenceInFirstRow(): void
+    {
+        $query = Customer::query();
+        $method = new ReflectionMethod(ActiveQuery::class, 'removeDuplicatedRows');
```
Comment (vjik) [inline tests/ActiveQueryTest.php]: Can we test it without reflection usage?

**Review by vjik — COMMENTED** [review]
Run cs-fixer and rector locally:
```shell
./vendor/bin/rector
./vendor/bin/php-cs-fixer fix
```

File: tests/ActiveQueryTest.php (hunk 33 above)
Comment (dbuhonov) [reply tests/ActiveQueryTest.php]: Done

**Comment by vjik** [comment]
@dbuhonov Fix CS, please

**Comment by vjik** [comment]
@dbuhonov Thank you!

**Comment by Tigrov** [comment]
@dbuhonov Great work! Several bugs were found 👍
